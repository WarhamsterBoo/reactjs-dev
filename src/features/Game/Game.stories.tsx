import React from "react";
import { Game } from ".";
import { Provider } from "react-redux";
import { store } from "@/AppStore";

export default { title: "Game component" };

export const GameStory: React.FC<{}> = () => {
  return (
    <Provider store={store}>
      <Game />
    </Provider>
  );
};
