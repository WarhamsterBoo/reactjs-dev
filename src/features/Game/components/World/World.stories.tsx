import React from "react";
import { World, CreatureState } from "./World";
import { withKnobs, object } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";

export default { title: "World component", decorators: [withKnobs] };

export const WorldStory: React.FC<{}> = () => {
  const creatures: CreatureState[][] = object("Creatures", [
    [{ isAlive: true }, { isAlive: true }, { isAlive: true }],
    [{ isAlive: true }, { isAlive: false }, { isAlive: false }],
    [{ isAlive: true }, { isAlive: true }, { isAlive: false }],
  ]);

  return <World creatures={creatures} onClick={action("button-click")} />;
};
