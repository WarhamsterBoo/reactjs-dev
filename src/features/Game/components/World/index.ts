export * from "./components";
export * from "./World";
