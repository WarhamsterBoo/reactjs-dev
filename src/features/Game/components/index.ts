export * from "./ControlPanel";
export * from "./SettingsForm";
export * from "./World";
export * from "./GameField";
