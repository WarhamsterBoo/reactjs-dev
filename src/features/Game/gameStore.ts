import { matrixGenerator, resizeMatrix } from "@/utils/arrayUtils";
import { AnyAction, createSlice, PayloadAction } from "@reduxjs/toolkit";

export interface WorldCreature {
  isAlive: boolean;
}

export interface GameSettings {
  xDimension: number;
  yDimension: number;
  fillingPercentage: number;
}

export interface GameState {
  settings: GameSettings;
  creatures: WorldCreature[][];
}

export interface CreatureCoordinates {
  x: number;
  y: number;
}

const initialState: GameState = {
  settings: { xDimension: 10, yDimension: 10, fillingPercentage: 0 },
  creatures: matrixGenerator(10, 10, { isAlive: false }),
};

const generateRandomCreatures = ({
  xDimension,
  yDimension,
  fillingPercentage,
}: GameSettings): WorldCreature[][] => {
  const creatures = matrixGenerator<WorldCreature>(xDimension, yDimension, {
    isAlive: false,
  });

  let NumberOfAliveCreatures = Math.trunc(
    xDimension * yDimension * fillingPercentage
  );
  while (NumberOfAliveCreatures > 0) {
    const x = Math.floor(Math.random() * Math.floor(xDimension));
    const y = Math.floor(Math.random() * Math.floor(yDimension));
    if (!creatures[x][y].isAlive) {
      creatures[x][y].isAlive = true;
      NumberOfAliveCreatures--;
    }
  }

  return creatures;
};

const changeCreaturesSize = (
  creatures: WorldCreature[][],
  xDimension: number,
  yDimension: number
): WorldCreature[][] => {
  return resizeMatrix(creatures, xDimension, yDimension, {
    isAlive: false,
  });
};

export const gameStore = createSlice({
  name: "game",
  initialState,
  reducers: {
    applySettings: (state, _: AnyAction) => state,
    saveSettings: (state, action: PayloadAction<GameSettings>) => {
      state.settings = action.payload;
    },
    resizeCreatures: (state, _: AnyAction) => {
      state.creatures = changeCreaturesSize(
        state.creatures,
        state.settings.xDimension,
        state.settings.yDimension
      );
    },
    generateNewCreatures: (state, _: AnyAction) => {
      state.creatures = generateRandomCreatures(state.settings);
    },
    changeSettingsTo: (state, _: PayloadAction<GameSettings>) => state,
    toggleCreatureState: (
      state,
      action: PayloadAction<CreatureCoordinates>
    ) => {
      const { x, y } = action.payload;
      state.creatures = state.creatures.map((row, rowIndex) => {
        if (rowIndex == x) {
          return row.map((value, inRowIndex) => ({
            isAlive: inRowIndex == y ? !value.isAlive : value.isAlive,
          }));
        }
        return row;
      });
    },
    stop: () => ({ ...initialState }),
  },
});
