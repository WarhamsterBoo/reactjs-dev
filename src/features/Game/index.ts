export * from "./components";
export * from "./Game";
export * from "./gameStore";
