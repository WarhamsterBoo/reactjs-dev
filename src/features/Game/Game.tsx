import { AppState } from "@/AppStore";
import React, { useCallback } from "react";
import { connect } from "react-redux";
import { ControlAction, ControlPanel, SettingsForm, World } from "./components";
import { GameWrapper } from "./Game.styled";
import {
  CreatureCoordinates,
  GameSettings,
  gameStore,
  Population,
} from "./gameStore";

interface GameProps {
  settings: GameSettings;
  onControlActionClick: (action: ControlAction) => void;
  creatures: Population;
  toggleCreatureState: (coordinates: CreatureCoordinates) => void;
  applySettings: () => void;
  onSettingsChange: (settings: GameSettings) => void;
}

const GameComponent: React.FC<GameProps> = ({
  settings,
  onControlActionClick,
  creatures,
  toggleCreatureState,
  applySettings,
  onSettingsChange,
}) => {
  const onCreatureClick = useCallback(
    (x: number, y: number) => toggleCreatureState({ x, y }),
    [toggleCreatureState]
  );

  return (
    <GameWrapper>
      <SettingsForm
        gameSettings={settings}
        onSettingsChange={onSettingsChange}
        applySettings={applySettings}
      />
      <World creatures={creatures} onClick={onCreatureClick} />
      <ControlPanel onControlButtonClick={onControlActionClick} />
    </GameWrapper>
  );
};

const mapStateToProps = (state: AppState) => {
  return {
    settings: state.game.settings,
    creatures: state.game.creatures,
  };
};

const mapDispatchToProps = {
  onControlActionClick: gameStore.actions.stop,
  toggleCreatureState: gameStore.actions.toggleCreatureState,
  applySettings: gameStore.actions.applySettings,
  onSettingsChange: gameStore.actions.changeSettings,
};

export const Game = connect(mapStateToProps, mapDispatchToProps)(GameComponent);
