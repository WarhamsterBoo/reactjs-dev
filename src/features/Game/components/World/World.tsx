import React from "react";
import { Creature } from "./components";
import { WorldWrapper } from "./World.styled";

export interface CreatureState {
  isAlive: boolean;
}

export interface WorldProps {
  creatures: CreatureState[][];
  onClick: (x: number, y: number) => void;
}

export const World: React.FC<WorldProps> = ({ creatures, onClick }) => {
  if (creatures.length == 0 || creatures[0].length == 0) {
    return null;
  }

  return (
    <WorldWrapper>
      {creatures.map((row, y) => [
        ...row.map((creature, x) => (
          <Creature
            key={`{${y}${x}}`}
            x={y}
            y={x}
            isAlive={creature.isAlive}
            onClick={onClick}
          />
        )),
        <br key={`${y}`} />,
      ])}
    </WorldWrapper>
  );
};
