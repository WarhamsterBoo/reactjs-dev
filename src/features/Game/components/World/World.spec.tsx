import { DEAD } from "@/features/Game/gameStore";
import { mount, shallow } from "enzyme";
import React from "react";
import { CreatureState, World } from "./World";

describe("World", () => {
  it.each`
    creatures
    ${[]}
    ${[[]]}
  `("should render nothing with 0 x 0 size: $creatures", ({ creatures }) => {
    const sut = shallow(<World creatures={creatures} onClick={jest.fn()} />);

    expect(sut).toMatchSnapshot();
  });

  it("should render with 1 x 1 size and creature is dead", () => {
    const creatures: CreatureState[][] = [[DEAD]];

    const sut = shallow(<World creatures={creatures} onClick={jest.fn()} />);

    expect(sut).toMatchSnapshot();
  });

  it("should render with 2 x 2 size filled with creatures", () => {
    const creatures: CreatureState[][] = [
      [DEAD, DEAD, DEAD],
      [DEAD, DEAD, DEAD],
      [DEAD, DEAD, DEAD],
    ];

    const sut = shallow(<World creatures={creatures} onClick={jest.fn()} />);

    expect(sut).toMatchSnapshot();
  });

  it("should call onClick callback with passed coordinates", () => {
    const creatures: CreatureState[][] = [[DEAD]];
    const fakeOnClick = jest.fn();
    const sut = mount(<World creatures={creatures} onClick={fakeOnClick} />);

    sut.find("button").simulate("click");

    expect(fakeOnClick).toBeCalledTimes(1);
    expect(fakeOnClick).toBeCalledWith(0, 0);
  });
});
