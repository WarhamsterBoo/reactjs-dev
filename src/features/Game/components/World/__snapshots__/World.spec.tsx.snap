// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`World should render nothing with 0 x 0 size: [[Array]] 1`] = `""`;

exports[`World should render nothing with 0 x 0 size: [] 1`] = `""`;

exports[`World should render with 1 x 1 size and creature is dead 1`] = `
<Styled(div)>
  <Creature
    isAlive={false}
    key="{00}"
    onClick={[MockFunction]}
    x={0}
    y={0}
  />
  <br
    key="0"
  />
</Styled(div)>
`;

exports[`World should render with 2 x 2 size filled with creatures 1`] = `
<Styled(div)>
  <Creature
    isAlive={false}
    key="{00}"
    onClick={[MockFunction]}
    x={0}
    y={0}
  />
  <Creature
    isAlive={false}
    key="{01}"
    onClick={[MockFunction]}
    x={0}
    y={1}
  />
  <Creature
    isAlive={false}
    key="{02}"
    onClick={[MockFunction]}
    x={0}
    y={2}
  />
  <br
    key="0"
  />
  <Creature
    isAlive={false}
    key="{10}"
    onClick={[MockFunction]}
    x={1}
    y={0}
  />
  <Creature
    isAlive={false}
    key="{11}"
    onClick={[MockFunction]}
    x={1}
    y={1}
  />
  <Creature
    isAlive={false}
    key="{12}"
    onClick={[MockFunction]}
    x={1}
    y={2}
  />
  <br
    key="1"
  />
  <Creature
    isAlive={false}
    key="{20}"
    onClick={[MockFunction]}
    x={2}
    y={0}
  />
  <Creature
    isAlive={false}
    key="{21}"
    onClick={[MockFunction]}
    x={2}
    y={1}
  />
  <Creature
    isAlive={false}
    key="{22}"
    onClick={[MockFunction]}
    x={2}
    y={2}
  />
  <br
    key="2"
  />
</Styled(div)>
`;
